// slack-client/lib/clients/rtm/client.js
/**
 *
 * The connection event sequence is:
 *   - CONNECTING             the rtm.start API call has been made
 *   - AUTHENTICATED          rtm.start returned successfully, with a websocket URL to connect to
 *   - OPENING_WEBSOCKET      the client is opening a socket at the supplied URL
 *   - OPENED_WEBSOCKET       the client has opened the socket at the supplied URL
 *   - OPENED_RTM_CONNECTION  the remote server has acked the socket and sent a `hello`, the connection is now live
 */

var bind = require('lodash').bind;
var contains = require('lodash').contains;
var inherits = require('inherits');
var isUndefined = require('lodash').isUndefined;

var RTM_EVENTS = require('./events/rtm-events').EVENTS;
var BaseAPIClient = require('../client');
var MemoryDataStore = require('../../data-store/memory-data-store');
var clientEvents = require('./events/client-events');
var makeMessageEventWithSubtype = require('./helpers').makeMessageEventWithSubtype;
var messageHandlers = require('./event-handlers');
var parseAPIResponse = require('../helpers').parseAPIResponse;
var wsSocketFn = require('../rtm/sockets/ws');


/**
 *
 * @param {object} webClient A Slack web client instance.
 * @param {object?} opts
 * @param {Function} opts.socketFn A function to call, passing in a websocket URL, that should return a websocket instance connected to that URL.
 * @param {object} opts.dataStore A store to cache Slack info, e.g. channels, users etc. in.
 * @param {boolean} opts.autoMark Whether to automatically mark messages as read after they're processed.
 * @param {boolean} opts.autoReconnect Whether or not to automatically reconnect when the connection closes.
 * @param {number} opts.maxReconnectionAttempts
 * @param {number} opts.reconnectionBackoff
 * @param {number} opts.wsPingInterval
 * @param {number} opts.maxPongInterval
 * @param {string} opts.logLevel The log level for the logger.
 * @param {Function} opts.logger Function to use for log calls, takes (logLevel, logString) parameters.
 * @constructor
 */
var RTMClient = function(webClient, opts) {

    opts = opts || {};

    BaseAPIClient.call(this, opts);

    /**
     * @private
     */
    this._webClient = webClient;

    /**
     * @type {Function}
     */
    this._socketFn = opts.socketFn || wsSocketFn;

    /**
     * The active websocket.
     * @type {Object}
     */
    this.ws;

    /**
     * @type {Object}
     */
    this.dataStore = opts.dataStore || new MemoryDataStore();

    this.MAX_RECONNECTION_ATTEMPTS = opts.maxReconnectionAttempts || 10;
    this.RECONNECTION_BACKOFF = opts.reconnectionBackoff || 3000;
    this.MAX_PONG_INTERVAL = opts.maxPongInterval || 10000;
    this.WS_PING_INTERVAL = opts.wsPingInterval || 5000;

    this.autoReconnect = opts.autoReconnect ? opts.autoReconnect : true;
};

inherits(RTMClient, BaseAPIClient);


/**
 * @type {boolean}
 */
RTMClient.prototype.connected = false;


/**
 * @type {boolean}
 */
RTMClient.prototype.authenticated = false;


/**
 * The id of the user that's currently connected via this client.
 * @type {string}
 */
RTMClient.prototype.activeUserId;


/**
 * The id of the team that's currently connected via this client.
 * @type {string}
 */
RTMClient.prototype.activeTeamId;


/**
 * Maps message id to message object for messages sent to but not replied to by the remote server.
 * @type {Object}
 * @private
 */
RTMClient.prototype._pendingMessages = {};


/**
 * The timer repeatedly pinging the server to let it know the client is still alive.
 * @type {?}
 */
RTMClient.prototype._pingTimer = null;


/**
 * The time the last pong was received from the server.
 * @type {number}
 * @private
 */
RTMClient.prototype._lastPong = 0;


/**
 *
 * @type {number}
 * @private
 */
RTMClient.prototype._connAttempts = 0;


/**
 * Whether the server is currently connecting.
 * @type {boolean}
 * @private
 */
RTMClient.prototype._connecting = false;


/**
 *
 * @param {object} opts
 */
RTMClient.prototype.start = function(opts) {
    // Check whether the client is currently attempting to connect to the RTM API.
    if (!this._connecting) {
        this.logger.log('verbose', 'attempting to connect via the RTM API');
        this.emit(clientEvents.CONNECTING);
        this._connecting = true;

        this._webClient.rtm.start(opts, bind(this._onStart, this));
    }
};


/**
 * @deprecated since version 2.0.0, use start() instead.
 */
RTMClient.prototype.login = function(opts) {
    this.start(opts);
};


/**
 * Generates the next message id to use.
 */
RTMClient.prototype.nextMessageId = function() {
    this._messageId = this._messageId || 1;
    return this._messageId++;
};


/**
 *
 * @param err
 * @param data
 * @private
 */
RTMClient.prototype._onStart = function(err, data) {
    this._connecting = false;
    if (err || !data['url']) {
        this.logger.log('verbose', 'failed to connect to the RTM API: ' + err);
        this.emit(clientEvents.FAILED_AUTHENTICATION, err);
        this.authenticated = false;
        if (this.autoReconnect) {
            this.reconnect();
        }
    } else {
        this.logger.log('verbose', 'rtm.start successful, attempting to open websocket URL');
        this.authenticated = true;
        this.activeUserId = data.self.id;
        this.activeTeamId = data.team.id;
        this.dataStore.cacheRtmStart(data);

        this.emit(clientEvents.AUTHENTICATED);
        this.connect(data['url']);
    }
};


/**
 * Closes the websocket and tears down the ping function.
 * @private
 */
RTMClient.prototype._safeDisconnect = function() {
    if (this._pingTimer) {
        clearInterval(this._pingTimer);
        this._pingTimer = null;
    }
    if (this.ws) {
        // Stop listening to the websocket's close event, so that the autoronnect logic doesn't fire twice.
        this.ws.removeAllListeners('close');
        this.ws.close();
    }
    this.authenticated = false;
    this.connected = false;
};


/**
 * Connects to the RTM API.
 * @param {string} socketUrl The URL of the websocket to connect to.
 */
RTMClient.prototype.connect = function(socketUrl) {
    this.emit(clientEvents.OPENING_WEBSOCKET);
    this.ws = this._socketFn(socketUrl);

    this.ws.on('open', bind(this.handleWsOpen, this));
    this.ws.on('message', bind(this.handleWsMessage, this));
    this.ws.on('error', bind(this.handleWsError, this));
    this.ws.on('close', bind(this.handleWsClose, this));
    this.ws.on('ping', bind(this.handleWsPing, this));
};


/**
 * Disconnects from the RTM API.
 */
RTMClient.prototype.disconnect = function() {
    this.autoReconnect = false;
    this._safeDisconnect();
};


/**
 *
 */
RTMClient.prototype.reconnect = function() {
    if (!this._connecting) {
        this.emit(clientEvents.ATTEMPTING_RECONNECT);
        this._safeDisconnect();

        this._connAttempts++;
        if (this._connAttempts > this.MAX_RECONNECTION_ATTEMPTS) {
            console.log('unable to connect to Slack RTM API, failed after max reconnection attempts for team ID: ' + this.teamId);
            this._safeDisconnect();
        }
        setTimeout(bind(this.start, this), this._connAttempts * this.RECONNECTION_BACKOFF);
    }
};


/**
 * Pings the remote server to let it know the client is still alive.
 * @private
 */
RTMClient.prototype._pingServer = function() {
    if (this.connected) {
        this.send({'type': 'ping'});

        // If the last pong was more than MAX_PONG_INTERVAL, force a reconnect
        var pongInterval = Date.now() - this._lastPong;
        if (pongInterval > this.MAX_PONG_INTERVAL) {
            this.reconnect();
        }
    }
};


/**
 * Handler to deal with the WebSocket open event.
 * NOTE: this.connected doesn't get set to true until the helloHandler is called.
 */
RTMClient.prototype.handleWsOpen = function() {
    this.emit(clientEvents.OPENED_WEBSOCKET);
    this._lastPong = Date.now();
    this._connAttempts = 0;
    if (this._pingTimer) {
        clearInterval(this._pingTimer);
    } else {
        this._pingTimer = setInterval(bind(this._pingServer, this), this.WS_PING_INTERVAL);
    }
};


/**
 * Handler to deal with the WebSocket message event.
 * @param {object} message
 */
RTMClient.prototype.handleWsMessage = function(message) {
    this.logger.log('debug', message);
    this.emit(clientEvents.RAW_MESSAGE, message);

    try {
        message = parseAPIResponse(message);
    } catch (err) {
        // TODO(leah): Emit an event here?
        this.logger.debug('unable to parse message: ' + err);
        return;
    }

    if (contains(['pong', RTM_EVENTS.HELLO, RTM_EVENTS.TEAM_MIGRATION_STARTED], message.type)) {
        this._handleWsMessageInternal(message.type, message);
    } else {
        this._handleWsMessageViaEventHandler(message.type, message);
    }
};


/**
 *
 * @param {String} messageType
 * @param {Object} message
 * @private
 */
RTMClient.prototype._handleWsMessageInternal = function(messageType, message) {
    if (messageType === 'pong') {
        this._handlePong(message);
    } else if (messageType === RTM_EVENTS.HELLO) {
        this._handleHello();
    } else if (messageType  === RTM_EVENTS.TEAM_MIGRATION_STARTED) {
        this._safeDisconnect();
        this.reconnect();
        this.emit(messageType, message);
    }
};


/**
 *
 * @param {String} messageType
 * @param {Object} message
 * @private
 */
RTMClient.prototype._handleWsMessageViaEventHandler = function(messageType, message) {
    if (messageType === RTM_EVENTS.MESSAGE && !isUndefined(message.subtype)) {
        var handler = messageHandlers[makeMessageEventWithSubtype(message.subtype)];
    } else {
        var handler = messageHandlers[messageType];
    }

    if (messageType === RTM_EVENTS.MESSAGE) {
        var replyTo = message.replyTo;
        if (replyTo) {
            // TODO(leah): figure out how message pushes via the RTM API should work and how any errors will be handled
            if (this._pendingMessages[replyTo]) {
                delete this._pendingMessages[replyTo];
            } else {
                // If the message was sent in reply to a message that's not on this client, skip the message
                return;
            }
        }

        // If there's a custom handler for the message subtype, use that, otherwise default to adding the
        // message to the base channel history
        handler = handler ? handler : messageHandlers[makeMessageEventWithSubtype('rtm_client_add_message')];
    }

    if (handler) {
        // Some handlers require the active user / team id, so support providing it as needed
        if (handler.length === 4) {
            handler(this.activeUserId, this.activeTeamId, this.dataStore, message);
        } else {
            handler(this.dataStore, message);
        }
    }

    this.emit(messageType, message);
    if (messageType === RTM_EVENTS.MESSAGE && !isUndefined(message.subtype)) {
        this.emit(makeMessageEventWithSubtype(message.subType), message);
    }
};


RTMClient.prototype.handleWsError = function() {
    //return this.emit('error', error);
};


/**
 *
 */
RTMClient.prototype.handleWsClose = function(code, reason) {
    this.connected = false;

    this.emit(clientEvents.WS_CLOSE, code, reason);

    if (this.autoReconnect && !this._connecting) {
        this.reconnect();
    }
};


/**
 * Handler for the websocket ping event.
 * This pongs back to the server to let it know the client is still active.
 */
RTMClient.prototype.handleWsPing = function() {
    if (this.ws.pong) {
        this.ws.pong();
    }
};


/**
 * Handles the server's pong message, updating the lastPong time on the client.
 * @param {Object} message
 */
RTMClient.prototype._handlePong = function(message) {
    this._lastPong = Date.now();
    delete this._pendingMessages[message.replyTo];
};


/** {@link https://api.slack.com/events/hello|hello} */
RTMClient.prototype._handleHello = function() {
    this.connected = true;
    this.emit(clientEvents.OPENED_RTM_CONNECTION);
};


/**
 * Sends a message over the websocket to the server.
 * @param {*} message The message to send back to the server.
 */
RTMClient.prototype.send = function(message) {
    if (this.connected) {
        message.id = this.nextMessageId();
        try {
            message = JSON.stringify(message);
            this._pendingMessages[message.id] = message;
            this.ws.send(message);
            return message;
        } catch (err) {
            // Debug log
        }
    }
};


module.exports = RTMClient;
